File: Version3.0/SimiCartPluginFW/SimiFramework/Model/SimiModel.m
//
//  SimiModel.m
//  SimiCartPluginFW
//
//  Created by Tan Hoang on 1/21/14.
//  Copyright (c) 2014 SimiTeam. All rights reserved.
//

#import "SimiModel.h"

@implementation SimiModel

+ (SimiAPI *)getSimiAPI
{
    NSString *klass = [NSStringFromClass(self) stringByReplacingOccurrencesOfString:@"Model" withString:@"API"];
    Class api = NSClassFromString(klass);
    Class loopClass = [self superclass];
    while (api == nil && loopClass) {
        api = NSClassFromString([[loopClass description] stringByReplacingOccurrencesOfString:@"Model" withString:@"API"]);
            loopClass = [loopClass superclass];
    }
    return [api new];
}

- (id)init
{
    self = [super init];
    if (self) {
        currentNotificationName = @"DidFinishRequest";
    }
    return self;
}

- (SimiAPI *)getAPI{
    return [self.class getSimiAPI];
}

- (void)setData:(NSDictionary *)data{
    if ([self count]) {
        [self removeAllObjects];
    }
    [self setValuesForKeysWithDictionary:data];
//    [self addEntriesFromDictionary:data];
}

- (void)addData:(NSDictionary *)data{
    [self addEntriesFromDictionary:data];
}

- (void)editData:(NSDictionary *)data{
    for (NSString *key in [data allKeys]) {
        if ([[self valueForKey:key] isKindOfClass:[NSMutableDictionary class]]) {
            NSMutableDictionary *origin = (NSMutableDictionary *)[self valueForKey:key];
            NSMutableDictionary *temp = (NSMutableDictionary *)[data valueForKey:key];
            for (NSString *childKey in [temp allKeys]) {
                [origin setValue:[temp valueForKey:childKey] forKey:childKey];
            }
        }else{
            [self setValue:[data valueForKey:key] forKey:key];
        }
    }
}

- (void)deleteData{
    [self removeAllObjects];
}

- (NSString *)currentNotificationName
{
    return currentNotificationName;
}

- (void)didFinishRequest:(NSObject *)responseObject responder:(SimiResponder *)responder{
    if (responder.simiObjectName) {
        currentNotificationName = responder.simiObjectName;
    }
     [[NSNotificationCenter defaultCenter]postNotificationName:@"TimeLoaderStop" object:currentNotificationName];
    if ([responseObject isKindOfClass:[SimiMutableDictionary class]]) {
        switch (modelActionType) {
            case ModelActionTypeInsert:{
                [self addData:(NSDictionary *)responseObject];
            }
                break;
            case ModelActionTypeEdit:{
                [self editData:(NSDictionary *)responseObject];
            }
                break;
            case ModelActionTypeDelete:{
                [self deleteData];
            }
                break;
            default:{
                [self setData:(NSDictionary *)responseObject];
            }
                break;
        }
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:currentNotificationName object:self userInfo:@{@"responder":responder}];
    if (![currentNotificationName isEqualToString:@"DidFinishRequest"]) {
        [[NSNotificationCenter defaultCenter] postNotificationName:@"DidFinishRequest" object:self userInfo:@{@"responder":responder}];
    }
}

- (void)preDoRequest{
    [[NSNotificationCenter defaultCenter] postNotificationName:[NSString stringWithFormat:@"%@Before", currentNotificationName] object:self];
    [[NSNotificationCenter defaultCenter]postNotificationName:@"TimeLoaderStart" object:currentNotificationName];
    if (![currentNotificationName isEqualToString:@"DidFinishRequest"]) {
        [[NSNotificationCenter defaultCenter] postNotificationName:@"DidFinishRequest-Before" object:self];
    }
}

@end

File: simitracking/simitracking/SimiFramework/Model/SimiModel.swift
//
//  SimiModel.swift
//  simitracking
//
//  Created by Hai Nguyen on 11/8/16.
//  Copyright © 2016 SimiCart. All rights reserved.
//

import Alamofire
import UIKit

class SimiModel: NSObject {
    
    static var simiAPI:SimiAPI!
    public var currentNotificationName:String = "DidFinishRequest"
    public var data:Dictionary<String, Any> = [:]
    
    public var error:Array<Dictionary<String, Any>>!
    public var isSucess = false
    
    
    public func getAPI()->SimiAPI{
        return SimiAPI()
    }
    
    public func deleteData() {
        data.removeAll()
    }
    
    public func setData(data: Dictionary<String, String>) {
        if (self.data.count>=1) {
            self.data.removeAll()
        }
        for (key, value) in data {
            self.data[key] = value
        }
    }
    
    public func addData(data: Dictionary<String, String>) {
        for (key, value) in data {
            self.data[key] = value
        }
    }
    
    public func getResource()->String {
        return ""
    }
    
    
    func preDoRequest() {
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName+"Before"), object: self)
        NotificationCenter.default.post(name: Notification.Name(rawValue: "TimeLoaderStart"), object: currentNotificationName)
        if (currentNotificationName != "DidFinishRequest") {
            NotificationCenter.default.post(name: Notification.Name(rawValue: "DidFinishRequest-Before"), object: self)
        }
    }
    
    public func didFinishRequest(responseObject:Dictionary<String, AnyObject>){
        self.error = nil
        if responseObject["errors"] != nil
        {
            self.error = responseObject["errors"] as! Array<Dictionary<String, Any>>
            self.isSucess = false
        }
        else {
            self.isSucess = true
        }
        if self.isSucess == true {
            self.data = responseObject[getResource()] as! Dictionary<String, Any>!
        }
        pushNotifications(responseObject: responseObject)
    }
    
    public func pushNotifications(responseObject:Dictionary<String, AnyObject>) {
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName+"After"), object: self)
        NotificationCenter.default.post(name: Notification.Name(rawValue: "TimeLoaderStop"), object: currentNotificationName)
        NotificationCenter.default.post(name: Notification.Name(rawValue: currentNotificationName), object: self, userInfo: responseObject)
        
        if (currentNotificationName != "DidFinishRequest") {
            NotificationCenter.default.post(name: Notification.Name(rawValue: "DidFinishRequest-After"), object: self)
        }
    }
}
